### PR TITLE
rpc: fix showing wrong amount when not all inputs are from me in gettransaction

### LINKED ISSUE #14136: rpc: Wrong `gettransaction` info for a coinjoin
Take a look at this transaction:

https://testnet.smartbit.com.au/tx/947ddb7d22b0c7e8edd3a56a92b8ec771aa8b18667ef508270f928b14821e924

I've double checked with `getaddressinfo` and only the first input is mine, and the second output is mine. So the transaction looks like

Inputs: [1.02 BTC mine,  1.1 BTC not-mine]
Outputs:  [0.10481803 BTC not-mine, 2.01 BTC mine]

So clearly this transaction makes me  0.99 BTC richer.  (-1.02 + 2.01)

So now let's look at what `gettransaction` returns:

```
{
  "amount": -0.10481803,
  "fee": 1.09481803,
  "confirmations": 0,
  "trusted": false,
  "txid": "947ddb7d22b0c7e8edd3a56a92b8ec771aa8b18667ef508270f928b14821e924",
  "walletconflicts": [
  ],
  "time": 1535929790,
  "timereceived": 1535929790,
  "bip125-replaceable": "no",
  "details": [
    {
      "address": "2N8qCRsKMJ3qNYo8LRb9kLpy3qiigAuedMT",
      "category": "send",
      "amount": -0.10481803,
      "vout": 0,
      "fee": 1.09481803,
      "abandoned": false
    },
    {
      "address": "2MzFcfPF56Jp6BC3FJk6yXwCY6FzhnbGt3o",
      "category": "send",
      "amount": -2.01000000,
      "label": "",
      "vout": 1,
      "fee": 1.09481803,
      "abandoned": false
    },
    {
      "address": "2MzFcfPF56Jp6BC3FJk6yXwCY6FzhnbGt3o",
      "category": "receive",
      "amount": 2.01000000,
      "label": "",
      "vout": 1
    }
  ],
  "hex": "020000000001022eb1b5698216f56eea3770c949a5d9555dcd1062571cc3113cae68132a5ba5660100000017160014f67a19d36a915d7e0360d3a66d1f055c7bd34687feffffffe165f072311e71825b47a4797221d7ae56d4b40b7707c540049aee43302448a401000000171600142fa1c410210f8228467e53f450c9e4e56337895ffeffffff028bf09f000000000017a914aaf6ba68cf6a2369445f119596adb84412d28d3a874004fb0b0000000017a9144cdbced31bab78d8a248204a961a4ad475d6b48a87024730440220724dbbdc278315ed367036b163749702c7810730dc2b17ba3226f72c7d0078ab02207945ffb8ffc78bcefa207ac8e172983b5c3d91f64c3cb869cdcd288e6c1d85610121037426b52590172ce72df80829e4a9f7530224306399c399080fc917d19ae44a72024730440220168c025d48f3ee4d626d37a85c92a75e778649e966d8f008caeed688cb6da73902202e00fb70c80e0ba591a12a8feb067cf3216eb054ac5059d0f38ce805bbdd478701210213be1638b4620ccacf4e1bec9748c3fc27561aa7169dbd1639be00bb20236ab800000000"
}
```

Note the "fee" thing it has is pretty confused. The actual fee is: 0.00518197 BTC

Running on the 0.17   branch if it makes a difference.

### PR DESCRIPTION
**Pull request by web3shannon**
Fix #14136.

In the rpc `gettransaction`, there is an error in calculating `nFee`.
```
CAmount nFee = (wtx.IsFromMe(filter) ? wtx.tx->GetValueOut() - nDebit : 0);
```

If not all the inputs are from me (like the case of CoinJoin), `wtx.IsFromMe(filter)` returns `true`, and `nFee` would get a wrong value from `wtx.tx->GetValueOut() - nDebit`. Here `nDebit` only stands for sum of my inputs rather than all inputs, so in this case, `nFee` would equal to all outputs minus my inputs.

A feasible solution is to use `IsAllFromMe` instead of `IsFromMe`. By this method, `nFee` is set as `0` and `amount` can be correctly calculated from `nCredit - nDebit`, aka, all my outputs minus all my inputs.

The right `amount` and `fee` in the case of #14136 is successfully got through this way, as illustrated below. Here we use `importaddress`, aka the watch-only mode, to simulate the case.

image

### CODE REVIEW

**Comment by promag** [comment]
Can you add a test to exercise this change?

**Comment by shannon1916** [comment]
Sure, one or two days later.

**Comment by shannon1916** [comment]
@promag Do you mean adding a `rpc_gettransaction.py` to `test/functional`?

**Comment by promag** [comment]
I think you can extend and existing test, there are some that use gettransaction.

**Comment by shannon1916** [comment]
The test is added.

File: test/functional/wallet_basic.py [hunk 1]
```diff
@@ -496,6 +496,26 @@ def run_test(self):
...
+        txid_0 = self.nodes[0].getblock(self.nodes[0].generatetoaddress(1, input_address_0)[0])['tx'][0]
+        txid_1 = self.nodes[0].getblock(self.nodes[0].generatetoaddress(1, input_address_1)[0])['tx'][0]
+        self.nodes[0].generate(100)
+        inputs = [{"txid": txid_0, "vout": 0}, {"txid": txid_1, "vout": 0}]
+        outputs = [{output_address_0: 49}, {output_address_1: 0.99}]
+        tx = self.nodes[0].createrawtransaction(inputs, outputs)
+        tx = self.nodes[0].signrawtransactionwithkey(tx, [input_privkey_0, input_privkey_1])['hex']
+        txid = self.nodes[0].sendrawtransaction(tx)
+        tx_obj = self.nodes[0].gettransaction(txid)
+        assert_equal(tx_obj['amount'], Decimal('24.00000000'))
```
Comment (promag) [inline test/functional/wallet_basic.py]: Could also assert keys `trusted = false`.

File: test/functional/wallet_basic.py [hunk 2]
```diff
@@ -496,6 +496,26 @@ def run_test(self):
...
+        # Build a tx where inputs = [25 mine, 25 not-mine] and outputs = [49 mine, 0.99 not-mine].
+        # The amount returned by gettransaction is expected to be 49 - 25 = 24.
+        self.nodes[0].generate(1)
+        input_address_0 = self.nodes[0].getnewaddress()
+        input_address_1 = self.nodes[1].getnewaddress()
+        input_privkey_0 = self.nodes[0].dumpprivkey(input_address_0)
+        input_privkey_1 = self.nodes[1].dumpprivkey(input_address_1)
+        output_address_0 = self.nodes[0].getnewaddress()
+        output_address_1 = self.nodes[1].getnewaddress()
+        txid_0 = self.nodes[0].getblock(self.nodes[0].generatetoaddress(1, input_address_0)[0])['tx'][0]
```
Comment (promag) [inline test/functional/wallet_basic.py]: At this point `self.nodes[1]` has some UTXO that can be used instead of generating more blocks.

**Review by promag — COMMENTED** [review]
This looks incomplete, please see `details` in `gettransaction` result - it shows fee 24.99.

File: test/functional/wallet_basic.py (hunk 2 above)
Comment (shannon1916) [reply test/functional/wallet_basic.py]: Obtaining txids and amounts of those existing UTXOs would make the code much more complex. Is it necessary?

**Comment by shannon1916** [comment]
@promag Yes, I'll fix the `details` part.

File: test/functional/wallet_basic.py (hunk 1 above)
Comment (shannon1916) [reply test/functional/wallet_basic.py]: OK. Will add more `assert_equal`s.

File: test/functional/wallet_basic.py (hunk 2 above)
Comment (promag) [reply test/functional/wallet_basic.py]: Just `listunspent`?

**Comment by shannon1916** [comment]
The `details` is fixed and the test is modified.

**Comment by instagibbs** [comment]
Before gathering more review I think this will need some concept ACKs.

For instance, I'm not sure returning a fee of 0 is the right thing here in cases of `~IsAllFromMe`?

**Comment by luke-jr** [comment]
I'm pretty sure it's impossible to do the right thing for CoinJoins here without extra metadata.

We can't just blindly add all the outputs as "sent", because we only sent some of them - but we have no way to know which one(s).

**Comment by adamjonas** [comment]
Hi @shannon1916 - It looks like this is still lacking concept ACKs. Is this something that you plan on continuing to pursue? #19002 and #18919 are also relevant and maybe you can piggyback off those approaches.

**Comment by shesek** [comment]
> I'm not sure returning a fee of 0 is the right thing here

I agree, I think it should be a `null` to differentiate from transactions that are actually paying no fee.

It would be great to see this issue get fixed.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1772,7 +1773,7 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
         COutputEntry output = {address, txout.nValue, (int)i};
 
         // If we are debited by the transaction, add the output as a "sent" entry
-        if (nDebit > 0)
+        if (fIsAllFromMe)
```
Comment (shesek) [inline src/wallet/wallet.cpp:1776]: Does this mean that a transaction where some inputs are wallet related and none of the outputs are won't get registered as either 'sent' or 'received'?

**Comment by MarcoFalke** [comment]
Closing for now. See https://github.com/bitcoin/bitcoin/issues/14136#issuecomment-1073884131

**Pull request closed without merging**